### DISCUSSION TITLE
MQTT state not updated by command

### DISCUSSION THREAD

**Discussion by MickPo** [HASwitchPlate/openHASP#579] [General]
So I've just spent the weekend designing pages for each place where I plan to place an openHASP device around my home... now for the easy fun part I thought.  Adding my prototype device (Sunton 2,8) to openHAB via MQTT.

I fired up MQTT-spy to take a look at the topics published for each change of the various buttons etc on each page.  I then followed the instructions on updating the status of buttons by publishing messages to the command topics.  What I have discovered is that I can view the state of openHASP objects e.g. hasp/xiao1/state/p2b1 {"event":"up","val":1} this seems fine.  I can publish messages to the command topic e.g. hasp/xiao1/command/p2b1.val.  0 for off and 1 for on.  This is reflected in the openHASP UI I have built.  GREAT!!!1!

Except.  These updates are never reflected in the status topic.  Unless I press the button on the screen the state value remains unchanged.  This would not be a problem if the openHASP UI were the only way to alter the state of an item such as a light, blind, etc however it is but one of many interfaces and needs to have its status updated to match changes to the objects which it is affecting.  Working with openHASP will become very confusing if the status does not reflect the state of objects in openHASP.

Please tell me I'm missing something?  I have read through the openHab examples posted here and they all seem to be about managing the openHASP device rather than controlling other devices via openHASP, which would seem to the main point of MQTT integration.  What am I missing?  Is this just a bug in the current nightly build which I'm running to support the Sunton hardware or am I just approaching things in the wrong way?

![image](https://github.com/HASwitchPlate/openHASP/assets/151452571/75232c75-fcc9-4667-ba41-92f811f332d7)

Above is an image of what I see in mqtt-spy.  I publish a message to haps/xiao1/command/p2b1.val and see the display update but do not see an update in the status.

![image](https://github.com/HASwitchPlate/openHASP/assets/151452571/3b91c8d7-33be-414c-8056-66970be0f48f)

Above: the switch is updated on the device but this isn't reflected in the state topic.

PS: I really love the UI I've been able to quickly build in openHASP and am excited by the prospect of adding one of these devices everywhere I currently have a switch.  I just need to work out how to make this work for me with openHAB.

**Comment by fvanroie** [comment]
If you send a message to `hasp/xiao1/command/p2b1.val` with an empty payload, then the state should be returned.
The idea is that when you change the state programatically, then the state is already known.
But you can still query the current state by leaving the payload empty and openHASP will reply the current state in the state topic.

**Reply by MickPo** [reply]
Thanks fvanroie.  I've been thinking about the process since posting (slept on it) and concluded that nothing would fall apart if the state topic is out of synch with the actual state so long as I never need to read it outside of an update.  I will just need to be a little circumspect with the triggers used in rules to ensure that I don't rely on the state being correct at other times.  

It's great to know that the state can be refreshed with an empty message to the command topic.  I can see this working in mqtt-spy now and think I could easily incorporate this into the rule for maintaining openHASP status synch.  

Thanks for helping me through this chicken little moment :). Using lvgl to drive these inexpensive displays vs the Nextion option is a great innovation IMHO.  I'll be sure to post pics to the openHab community once I've implemented my openHASP everywhere vision.
